source/Cosmos.HAL2/SoundBlaster16.cs

```diff
@@ -0,0 +1,324 @@
+            SampleRate = (byte)(constt >> 8);
+
+
+            //Program 8-bit transfers
+            Outb(0x0A, 0x05); //Disable channel 1 (Channel # + 0x04)
+            Outb(0x0C, 0); //Flip flop flip port
+            Outb(0x0B, 0x59); //Auto Init mode
+
+            //Send sound data location
+            Outb(0x83, firstAudioPosition); //Page # example: 0x[01]0F04
+            Outb(0x02, lowerAudioPosition); //Lower audio position bits  0x01[0F]04
+            Outb(0x02, upperAudioPosition); //high audio position bits 0x010F[04]
+
+            //Send length of data
+            Outb(0x03, lowerAudioLength); //Send low bits of audio length
+            Outb(0x03, highAudioLength); //Send high bits of audio length
+            Outb(0x0A, 1); //Enable channel 1
+
+            //Set time constant (speed)
+            WriteDSP(DSP_SETTIME);
+            WriteDSP(SampleRate);
+
+            //set block size for 8-bit auto-init mode
+            WriteDSP(0x48);
+
+            //Send data length to DSP
+            WriteDSP((byte)(lowerAudioLength - 1));
+            WriteDSP((byte)(highAudioLength - 1));
+
+            //Start auto init 8-bit transfer
+            WriteDSP(0x1c);
+        }
+        private void IrqHandler(ref INTs.IRQContext c)
+        {
+            Global.mDebugger.Send("Got Sound blaster 16 IRQ");
+            PlayingSound = false;
+
+            //Play the next sound in the buffer
+            if (IsAudioBiggerThenBuffer)
+            {
+                IsAudioBiggerThenBuffer = false;
+
+                //Clear buffer
+                for (uint i = 0; i < Buffer.Length; i++)
+                {
+                    Buffer[i] = 0;
+                    bufferinmem[i] = 0;
+                }
+
+                //Copy the next audio data to the buffer
+                for (int i = AudioStop; i < audiodata.Length; i++)
+                {
+                    if (i >= Buffer.Length)
+                    {
+                        Global.mDebugger.Send("Audio Buffer " + i);
+                        IsAudioBiggerThenBuffer = true;
+                        AudioStop = i;
+                        break;
+                    }
+                    else
+                    {
+                        Buffer[i] = audiodata[i];
+                    }
+                }
+
+                //Play the buffer
+                PlayBuffer();
+            }
+            else
+            {
+                //Clean up
+                PlayingSound = false;
+                IsAudioBiggerThenBuffer = false;
+                AudioStop = 0;
+                audiodata = null;
+            }
+        }
+        #region I/O
+        private void MixerPortWrite(byte port, byte data)
+        {
+            Outb(MixerPort, port);
+            Outb(MixerDataPort, data);
+        }
+        private byte MixerPortRead()
+        {
+            return Inb(MixerDataPort);
+        }
+        private static byte Inb(ushort port)
```

Comment:
Oh, I also added an inb and outb function for another pull request: https://github.com/CosmosOS/Cosmos/blob/rtl8168/source/Cosmos.Core/Ports.cs
@quajak do you think that every device needs an IOGroup class like the PCNETII driver Cosmos.Core/IOGroup/Network/AMDPCNetIIIOGroup.cs), or we can just add a class for IO Ports with functions like inb, outb and so on?

Reply:
I think it should be fine if we just use the `Port` directly in the driver class. Or is there some advantage to IOGRoups I am not thinking of?